test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java

```diff
@@ -0,0 +1,188 @@
+        String fullTableName = KEYSPACE + '.' + tableName;
+        cluster.schemaChange(withKeyspace("CREATE TABLE %s." + tableName + " (k int PRIMARY KEY, v text)"));
+
+        // The failure threshold provided is exactly the total number of rows returned
+        // to the coordinator from all replicas and therefore should just warn.
+        testMissedUpdates(fullTableName, 1, REPLICAS * ROWS, true);
+
+        try
+        {
+            // The failure threshold provided is one less than the total number of rows returned
+            // to the coordinator from all replicas and therefore should fail the query.
+            testMissedUpdates(fullTableName, 1, REPLICAS * ROWS - 1, true);
+        }
+        catch (RuntimeException e)
+        {
+            assertEquals(e.getCause().getClass().getName(), TooManyCachedRowsException.class.getName());
+        }
+    }
+
+    private void testMissedUpdates(String fullTableName, int warnThreshold, int failThreshold, boolean shouldWarn)
+    {
+        cluster.get(1).runOnInstance(() -> StorageService.instance.setCachedReplicaRowsWarnThreshold(warnThreshold));
+        cluster.get(1).runOnInstance(() -> StorageService.instance.setCachedReplicaRowsFailThreshold(failThreshold));
+
+        for (int i = 0; i < ROWS; i++)
+        {
+            cluster.coordinator(1).execute("INSERT INTO " + fullTableName + "(k, v) VALUES (?, 'old')", ALL, i);
+        }
+
+        String query = "SELECT * FROM " + fullTableName + " WHERE v = ? LIMIT ? ALLOW FILTERING";
+
+        Object[][] initialRows = cluster.coordinator(1).execute(query, ALL, "old", ROWS);
+        assertRows(initialRows, row(1, "old"), row(0, "old"), row(2, "old"));
+
+        // Update all rows on only one replica, leaving the entire dataset of the remaining replica out-of-date:
+        for (int i = 0; i < ROWS; i++)
+        {
+            cluster.get(1).executeInternal("UPDATE " + fullTableName + " SET v = 'new' WHERE k = ?", i);
+        }
+
+        // TODO: These should be able to use ICoordinator#executeWithResult() once CASSANDRA-15920 is resolved.
+        Object[] oldResponse = cluster.get(1).callOnInstance(() -> executeInternal(query, "old", ROWS));
+        Object[][] oldRows = (Object[][]) oldResponse[0];
+        assertRows(oldRows);
+        @SuppressWarnings("unchecked") List<String> oldWarnings = (List<String>) oldResponse[1];
+        assertEquals(shouldWarn, oldWarnings.stream().anyMatch(w -> w.contains("cached_replica_rows_warn_threshold")));
+
+        Object[] newResponse = cluster.get(1).callOnInstance(() -> executeInternal(query, "new", ROWS));
+        Object[][] newRows = (Object[][]) newResponse[0];
+        assertRows(newRows, row(1, "new"), row(0, "new"), row(2, "new"));
+        @SuppressWarnings("unchecked") List<String> newWarnings = (List<String>) newResponse[1];
+        assertEquals(shouldWarn, newWarnings.stream().anyMatch(w -> w.contains("cached_replica_rows_warn_threshold")));
```

Comment:
Could we check that it only warns once per query?

Reply:
It seems we are missing this nit in the last commit.

Reply:
@adelapena Forgot to push my last local commit ;)